visualqa/trainMLP.py

```diff
@@ -95,33 +106,81 @@ def main():
     data_root = args.dataroot
     experiment_root = args.experiment_root
 
+    # Set both the numpy and the Python random seeds.
+    random.seed(RANDOM_SEED)
+    np.random.seed(RANDOM_SEED)
+
     # Load data needed for training and save all parameters/mappings to make
     # sure experiments are reproducible
-    questions_train, answers_train, images_train = load_train_data(data_root)
+    questions_train_all, answers_train_all, images_train_all = load_train_data(
+        data_root)
 
     # Since we are simplifying the problem of Visual QA to a classification
     # problem in this baseline, we want to limit the number of possible
     # answers, and have the model simply pick the most appropriate one.
     max_answers = 1000
-    questions_train, answers_train, images_train = select_frequent_answers(
-        questions_train, answers_train, images_train, max_answers)
+    questions_train_all, answers_train_all, images_train_all = \
+        select_frequent_answers(questions_train_all, answers_train_all,
+                                images_train_all, max_answers)
 
     # Encode the remaining (top max_answers) answers and save the mapping.
     labelencoder = preprocessing.LabelEncoder()
-    labelencoder.fit(answers_train)
+    labelencoder.fit(answers_train_all)
     nb_classes = len(list(labelencoder.classes_))
     with open(pjoin(experiment_root, 'labelencoder.pkl'), 'wb') as pfile:
         pickle.dump(labelencoder, pfile)
 
+    # The initial shuffle ensures that the train-val split is randomized
+    # depending on the random seed, and not fixed every time (which would be
+    # very bad).
+    print("Performing initial shuffle...")
+    questions_train_all, answers_train_all, images_train_all = sklearn_shuffle(
+        questions_train_all, answers_train_all, images_train_all
+    )
+
+    train_all_count = len(questions_train_all)
+    valid_count = int(train_all_count * args.valid_ratio)
+    train_count = train_all_count - valid_count
+
+    print("We have {0} total Q-A pairs. Will use {1:.2f}% for validation, "
+          "which is {2} data points. {3} data points will be used for "
+          "actual training.".format(train_all_count, args.valid_ratio * 100.0,
+                                    valid_count, train_count))
+
+    questions_train = questions_train_all[:train_count]
```

Comment:
Are you sure questions are not ordered in any kind like yes/no before numbers? In that case we would need to permute everything to really select a random subset.